### PR TITLE
fix: OneToManyPersister does not take custom identifier types into account for orphan removal

### PR DESCRIPTION
**Pull request by wtfzdotnet**
I've taken a look at the test suite, but I am kind of mind-boggled as of where to begin, could you please point me in the right direction to make sure this change gets merged? (I'll tidy up this PR later when it's done)

Thanks! :-)

### CODE REVIEW

**Comment by mpdude** [comment]
Maybe the functional tests named `GHxxxTest` according to issue numbers give you enough inspiration to get started?

**Comment by wtfzdotnet** [comment]
> Maybe the functional tests named `GHxxxTest` according to issue numbers give you enough inspiration to get started?

Yeah I would've just placed this next to the existing ManyToManyPersister test, so good I asked, thanks :-)

**Comment by wtfzdotnet** [comment]
Okay I've made a unit test that validates this behavior ( tested on mysql, and it fails as expected ), adding Ramsey as a vendor obviously doesn't sense, but I could use a little help with ideas as of how to go about testing this via another route :-).

File: composer.json
Comment (greg0ire) [inline composer.json]: Other route: create a custom type in the test: https://www.doctrine-project.org/projects/doctrine-dbal/en/current/reference/types.html#custom-mapping-types

Comment (wtfzdotnet) [reply composer.json]: I did that in the first place, but SQLite couldn't be bothered in memory with my assertions for some reason. After I've set up a docker MySQL instance I got further, I'll try this again later with a simplistic example. Heading out to enjoy the sun for a bit now :-)

Comment (greg0ire) [reply composer.json]: > SQLite couldn't be bothered in memory with my assertions for some reason

I didn't quite get that sentence, but enjoy the :sun_with_face: !

Comment (wtfzdotnet) [reply composer.json]: Oh, when I ran the test suite early just with the default SQLite in memory configuration i couldn't get a failing assertion to my surprise. But I was also asserting against what's still left over in the unit of work instead of really querying manually again, so i think that's what caused confusion.

**Comment by wtfzdotnet** [comment]
@greg0ire think we are good now for a review :-)

File: composer.json
Comment (wtfzdotnet) [reply composer.json]: SQLite also failed with manual querying so I think that's where I made a wrong assumption :-)

**Comment by wtfzdotnet** [comment]
Also fixes #9745

File: tests/Doctrine/Tests/ORM/Functional/GH10747Test.php
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/GH10747Test.php]: Let's be consistent, please.

```suggestion
     * @var Collection<int, GH10747Credit>
```

**Comment by greg0ire** [comment]
> think we are good now for a review 

There are PHPStan issues. See https://www.doctrine-project.org/contribute/#running-checks-before-pushing for how to run checks before pushing.

**Comment by wtfzdotnet** [comment]
> > think we are good now for a review
> 
> There are PHPStan issues. See https://www.doctrine-project.org/contribute/#running-checks-before-pushing for how to run checks before pushing.

I ran this earlier locally, and again after your comment ( I did mess up phpcs, and psalm apparently ).

```
➜  orm git:(feature/fix-one-to-many-custom-id-orphan-removal) vendor/bin/phpstan
Note: Using configuration file /home/m/git/doctrine/orm/phpstan.neon.
 454/454 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%


                                                                                                                        
 [OK] No errors                                                                                                         
```

PHPCS issue is fixed, inconsistency is fixed, but I cannot grasp how the PHPStan errors from the report are related to my PR, could you please double check if I am a dummy?

As for the psalm error, it seems related to the underlying existing documentation which already supports `$types`, most of the code is just borrowed from the `ManyToManyPersister` to begin with so I am not sure what I'd have to fix here, I am not so familiar with psalm (unfortunately).

**Comment by greg0ire** [comment]
> PHPCS issue is fixed, incosistency is fixed, but I cannot grasp how the PHPStan errors from the report are related to my PR, could you please double check if I am a dummy?

Ah I didn't check myself, maybe it's related to a recent release, let me check.

**Comment by greg0ire** [comment]
Some of it is definitely related to a recent release, I'll fix it.

**Comment by wtfzdotnet** [comment]
Thanks let me know when I should rebase, and if I'd have to do anything else in regards to the psalm error :-), and thanks for your prompt feedback!

**Comment by greg0ire** [comment]
Ok so you can ignore the second Psalm error, and focus on the 2 other ones, they must be caused by this. See #10754 for the other errors.

**Comment by wtfzdotnet** [comment]
> 

You'd want me to just rebase on top of that branch or wait until it merges?

**Comment by wtfzdotnet** [comment]
@greg0ire as I said I am really not familiar with all the ins and outs of psalm, I ran into these baseline errors, after some searching around I ran the `vendor/bin/psalm --update-baseline`. ( Assuming the baseline is ignoring stuff now that is irrelevant, which cause the errors? )



> Ok so you can ignore the second Psalm error, and focus on the 2 other ones, they must be caused by this. See #10754 for the other errors.

I updated the psalm baseline and committed this, assuming that's what is supposed to happen at this stage. As far as I understand it's a sort of "ignore stuff from here", and the errors triggered seem related. I might be misunderstanding this so if I need to do something else please give me a little more guidance, psalm noob :-).

I could imagine this might have more changes than it should, but I am not familiar as of how to treat this file.

**Comment by greg0ire** [comment]
You will need to wait for #10754 to be merged. The baseline is a file that allows you to ignore errors. Setting Psalm to a very high level generates many errors, too many for us, so we baselined many of them. That way, new code is held to a very high standard, and we can reduce the baseline progressively and make old code also comply with high standards. In your case, it's OK to remove stuff from the baseline if your changes somehow fix them, or make them no longer relevant.

File: psalm-baseline.xml
Comment (greg0ire) [inline psalm-baseline.xml]: You shouldn't touch sections for files you didn't modify, that's for sure. I suppose that was done automatically, so maybe you don't have the correct version of Psalm installed? Make sure to run `composer update`.

Comment (wtfzdotnet) [reply psalm-baseline.xml]: Yeah so basically I'd have to go over the change of the baseline file and make sure it only fixes stuff in-scope of my pr, gotcha. I'll sort it out.

**Comment by wtfzdotnet** [comment]
> You will need to wait for #10754 to be merged. The baseline is a file that allows you to ignore errors. Setting Psalm to a very high level generates many errors, too many for us, so we baselined many of them. That way, new code is held to a very high standard, and we can reduce the baseline progressively and make old code also comply with high standards. In your case, it's OK to remove stuff from the baseline if your changes somehow fix them, or make them no longer relevant.

Assumed that was the case, thanks for taking the time to explain :-)

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php
Comment (wtfzdotnet) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: psalm kept whining about the return type either being `int|string`, given php version i changed these into `mixed`

Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: That won't work on old versions of PHP

Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: I think the right way to deal with this is to store the result in a variable, then assert it's an integer, and not touch the return type.

Comment (wtfzdotnet) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: Will fix it after lunch break, thanks.

Comment (wtfzdotnet) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: @greg0ire got distracted for a bit there :-), making assertions on the integers doesn't satisfy psalm, so I think this should simply be ignored in either the baseline or another way?

Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: Can't see why it wouldn't satisfy Psalm. Can you show me what you tried, and what's the result, by pushing your attempt to this branch?

Comment (wtfzdotnet) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: @greg0ire assumed this is what you meant, which results in

```
ERROR: MissingReturnType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:169:22 - Method Doctrine\ORM\Persisters\Collection\OneToManyPersister::deleteEntityCollection does not have a return type, expecting int|string (see https://psalm.dev/050)
    private function deleteEntityCollection(PersistentCollection $collection)


ERROR: RedundantConditionGivenDocblockType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:190:40 - Docblock-defined type string for $numAffected is always string (see https://psalm.dev/156)
        assert(is_int($numAffected) || is_string($numAffected));


ERROR: MissingReturnType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:203:22 - Method Doctrine\ORM\Persisters\Collection\OneToManyPersister::deleteJoinedEntityCollection does not have a return type, expecting int|string (see https://psalm.dev/050)
    private function deleteJoinedEntityCollection(PersistentCollection $collection)


ERROR: RedundantConditionGivenDocblockType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:256:39 - Docblock-defined type string for $numDeleted is always string (see https://psalm.dev/156)
        assert(is_int($numDeleted) || is_string($numDeleted));
```

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: I meant this:

```suggestion
        assert(is_int($numAffected));
```

In that case, you will have to ignore an error about the possibility of `$numAffected` being a string, in case it's gigantic I guess.

File: psalm-baseline.xml
Comment (greg0ire) [inline psalm-baseline.xml:1246]: This should probably be kept once you've addressed my comment above. Part of this, at least.

Comment (wtfzdotnet) [reply psalm-baseline.xml:1246]: @greg0ire at that stage ( just resetting the InvalidReturnType node ), I am left with;

```
ERROR: MissingReturnType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:169:22 - Method Doctrine\ORM\Persisters\Collection\OneToManyPersister::deleteEntityCollection does not have a return type, expecting int (see https://psalm.dev/050)
    private function deleteEntityCollection(PersistentCollection $collection)


ERROR: MissingReturnType - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:203:22 - Method Doctrine\ORM\Persisters\Collection\OneToManyPersister::deleteJoinedEntityCollection does not have a return type, expecting int (see https://psalm.dev/050)
    private function deleteJoinedEntityCollection(PersistentCollection $collection)


ERROR: UnusedBaselineEntry - lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:0:0 - Baseline for issue "InvalidReturnType" has 2 extra entries. (see https://psalm.dev/316)
```

Comment (greg0ire) [reply psalm-baseline.xml:1246]: Well now that you asserted that it's an integer, you should add a return type declaration at least to `deleteEntityCollection`, right? Same for the other method.

Comment (wtfzdotnet) [reply psalm-baseline.xml:1246]: Ok that made me feel like a clown, I was presuming it was nagging about the return type hints again

**Comment by wtfzdotnet** [comment]
phpstan happy, psalm happy, phpcs happy, I hope we are finally good now :pray:

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: :thinking: I don't understand why this was not reverted.

Comment (wtfzdotnet) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: The @throws  ?

Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: The removal of the type declaration.

Comment (wtfzdotnet) [reply lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: Apologies, really haven't been having my day, bad night of sleep :sleeping:

**Comment by wtfzdotnet** [comment]
Let me know if I still need to squash these commits, sorry about the going back and forth :heart:

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php:257]: Did you test this locally?

**Comment by greg0ire** [comment]
> Let me know if I still need to squash these commits, sorry about the going back and forth heart

No worries. I think there should be 2 commits:
- one commit that adds the assertions and changes the baseline
- the commit with the changes you wanted to add

That's because right now, I don't think there is a relation between the 2 changes. Let me know if you need help going from 3 to 2 commits.

**Comment by wtfzdotnet** [comment]
> > Let me know if I still need to squash these commits, sorry about the going back and forth heart
> 
> No worries. I think there should be 2 commits:
> 
> * one commit that adds the assertions and changes the baseline
> * the commit with the changes you wanted to add
> 
> That's because right now, I don't think there is a relation between the 2 changes. Let me know if you need help going from 3 to 2 commits.

@greg0ire okay that's done as well :sweat:

**Comment by greg0ire** [comment]
The code looks good, I think. The commit message should be improved, so you should probably read https://www.doctrine-project.org/contribute/#crafting-meaningful-commit-messages

Specifically, the commit message only mentions the problem you're addressing, but does not detail what happens (do you have an error message? Or nothing happens. Also, it does not explain why why the issue exists, what you are doing in this PR, and why that fixes the issue.

Maybe what's happening is that at some point, the ORM is comparing objects with their representation in the database, when it should really convert the objects first, is that it?

**Comment by wtfzdotnet** [comment]
> The code looks good, I think. The commit message should be improved, so you should probably read https://www.doctrine-project.org/contribute/#crafting-meaningful-commit-messages
> 
> Specifically, the commit message only mentions the problem you're addressing, but does not detail what happens (do you have an error message? Or nothing happens. Also, it does not explain why why the issue exists, what you are doing in this PR, and why that fixes the issue.
> 
> Maybe what's happening is that at some point, the ORM is comparing objects with their representation in the database, when it should really convert the objects first, is that it?

As was in the example I've given before with the gnarly ramsey/uuid implementation ( for the test case ), basically what happens is a custom doctrine type of Uuid object is converted to string by simply casting, resulting in a hex `DELETE FROM x WHERE id = ?` query, whilst it should've been converted along the way to it's  binary representation ( in our database choice case ). This leads to no deletions being made at all as you would expect making use of doctrine custom type's as an identifier.

**Comment by wtfzdotnet** [comment]
Rewrote last commit's message to address this.

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Comment by norkunas** [comment]
And #10075 just was ignored, thanks guys

**Comment by greg0ire** [comment]
Thanks @wtfzdotnet !

@norkunas , that's unfortunate your PR got ignored, but I don't see why you feel entitled to complain about it. I mean, we're not paid to review all incoming PRs, are we?

**Comment by norkunas** [comment]
I don't complain. I stated a fact, and then thanked. At least after so long somebody gave a duck and completed this

**Comment by greg0ire** [comment]
If you don't complain then we're cool :slightly_smiling_face: We in the team get drowned daily in a shitload of notifications, and it's indeed cool that people don't get discouraged by this and stuff ends up getting fixed. So thank you both for caring!

**Pull request merged**
